src/wp-includes/html-api/class-wp-css-selectors.php

```diff
@@ -0,0 +1,820 @@
+<?php
+/**
+ * HTML API: WP_CSS_Selectors class
+ *
+ * @package WordPress
+ * @subpackage HTML-API
+ * @since TBD
+ */
+
+// phpcs:disable Generic.Files.OneObjectStructurePerFile.MultipleFound
+
+/**
+ * Core class used by the HTML processor to parse CSS selectors.
+ *
+ * This class is designed for internal use by the HTML processor.
+ *
+ * This class is instantiated via the `WP_CSS_Selector::from_selector( string $selector )` method.
+ * It accepts a CSS selector string and returns an instance of itself or `null` if the selector
+ * is invalid or unsupported.
+ *
+ * A subset of the CSS selector grammar is supported. The grammar is defined in the CSS Syntax
+ * specification, which is available at https://www.w3.org/TR/css-syntax-3/.
+ *
+ * Supported selector syntax:
+ * - Type selectors (tag names, e.g. `div`)
+ * - Class selectors (e.g. `.class-name`)
+ * - ID selectors (e.g. `#unique-id`)
+ * - Attribute selectors (e.g. `[attribute-name]` or `[attribute-name="value"]`)
+ * - Comma-separated selector lists (e.g. `.selector-1, .selector-2`)
+ * - The following combinators:
+ *   - descendant (e.g. `.parent .descendant`)
+ *   - child (`.parent > .child`)
```

Comment:
I don't think this will be supported initially. Maybe no combinators at all, only simple selectors.

These two combinators can probably be supported by the HTML API as long as they non-final selector is an element selector:

- `div > .className` ✅ supported
- `.className > div` ⛔️ not supported

Tags can be "seen" via bookmarks, while things like IDs, classes, attributes etc. would require seeking or more advanced internal tracking.

Reply:
Is the problem that we'd need to keep track of all the attributes requested by the selector in the entire breadcrumbs trail?

Reply:
Exactly. Either breadcrumbs would need to store all of their attributes or we'd need to seek around the document to check them which would be very costly.

All of the supported selectors are things that can be checked from given tag in the document without any seeking as the HTML API is implemented now.

Reply:
If we knew the selector upfront, we could always store the attributes needed to match it, e.g. store the `class` attribute if we know we'll only ever look for `div > .className`. Only that wouldn't be practical.

However, keeping track of classnames in all the breadcrumbs elements seems okay-ish resource-wise.

Perhaps we could even have a "CSS-enabled" mode where we'd keep track of all the breadcrumbs attributes shorter than, say, 100 bytes? It wouldn't be perfect but should suffice for most attribute-based matches.

Reply:
There will be ways to support more selectors, one good idea is to set some state to track data for certain selectors as the document is traversed.

One tricky thing with that implementation is that it may be necessary to return to the start of the document depending on the selectors used and scan again from there so that the necessary data is collected during traversal. This probably involves analyzing the selectors, deciding what data needs to be stored, seeking to the beginning if necessary, seeking back to the current location, and then trying to find the selector.

The complexity here is significant. It can be overcome, but I think the set of selectors supported in this PR as it is now is a good starting place with very limited matching complexity. Iteration to improve support can be added in future PRs.

Reply:
Oh! Everything I said I meant for the future, no need to hold up this work :) 

On seeking - sounds reasonable, although it won't combine well with streaming. What would an example of a selector that requires backtracking?

Reply:
> What would an example of a selector that requires backtracking?

This is mostly necessary to support selecting from arbitrary positions in the document. If we knew we were at the start of the document, we could determine what additional state needed to be tracked and start collecting it.

- The sibling combinators `+` and `~` select based on preceding elements, so it would be necessary to back up to check for matches.
- Support for subclass selectors (ID, class, and attribute) in non-final position, like `#my-id div` would require backing up to parse some attributes along with breadcrumbs.
- If pseudo-class selectors were added, I'd expect many of them to require more understanding of the document structure.